oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/repository/RepositoryImpl.java

```diff
@@ -564,4 +571,30 @@ public synchronized void cancel() {
             }
         }
     }
+
+    /**
+     * This is a fallback implementation of {@link org.apache.jackrabbit.oak.spi.query.SessionQuerySettingsProvider} that
+     * unifies and replaces the previous handling of {@code fastQueryResultSize} and {@code oak.fastQuerySize} here
+     * and in the {@link org.apache.jackrabbit.oak.jcr.session.SessionContext}.
+     */
+    private static class FastQuerySizeSettingsProvider implements SessionQuerySettingsProvider {
+        private final boolean fastQueryResultSize;
+
+        public FastQuerySizeSettingsProvider(boolean fastQueryResultSize) {
+            this.fastQueryResultSize = fastQueryResultSize;
+        }
+
+        @Override
+        public @NotNull SessionQuerySettings getQuerySettings(@NotNull ContentSession session) {
+            return new SessionQuerySettings() {
+                @Override
+                public boolean useDirectResultCount() {
+                    if (System.getProperty("oak.fastQuerySize") != null) {
+                        return Boolean.getBoolean("oak.fastQuerySize");
+                    }
```

Comment:
I would access the system property only once... Something like this:
```suggestion
                    String x = System.getProperty("oak.fastQuerySize");
                    if (x != null) {
                        return Boolean.parseBoolean(x);
                    }
```

I know the risk of concurrent change to the system property is tiny... but still.

Reply:
@thomasmueller Thanks. I made a change to use an Optional expression for the line.

Reply:
I personally would have preferred the "if" statement (it's shorter) but OK.